x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/tree/NodeSubclassTests.java

```diff
@@ -618,45 +622,77 @@ protected boolean hasAtLeastTwoChildren(Class<? extends Node<?>> toBuildClass) {
-                        }
 
-                        if (false == Modifier.isAbstract(c.getModifiers())
-                                && false == c.isAnonymousClass()
-                                && clazz.isAssignableFrom(c)) {
-                            Class<? extends T> s = c.asSubclass(clazz);
-                            results.add(s);
+            // load classes from jar files
+            // NIO FileSystem API is not used since it trips the SecurityManager
+            // https://bugs.openjdk.java.net/browse/JDK-8160798
+            // so iterate the jar "by hand"
+            if (path.endsWith(".jar") && path.contains("x-pack-ql")) {
+                try (JarInputStream jar = jarStream(root)) {
+                    JarEntry je = null;
+                    while ((je = jar.getNextJarEntry()) != null) {
+                        String name = je.getName();
+                        if (name.endsWith(".class")) {
+                            String className = name.substring(0, name.length() - ".class".length()).replace("/", ".");
+                            maybeLoadClass(clazz, className, root + "!/" + name, results);
                         }
                     }
-                    return FileVisitResult.CONTINUE;
                 }
-            });
+            }
+            // for folders, just use the FileSystems API
+            else {
+                Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+                    @Override
+                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                        if (Files.isRegularFile(file) && file.getFileName().toString().endsWith(".class")) {
+                            String fileName = file.toString();
+                            // Chop off the root and file extension
+                            String className = fileName.substring(rootLength, fileName.length() - ".class".length());
+                            // Go from "path" style to class style
+                            className = className.replace(PathUtils.getDefaultFileSystem().getSeparator(), ".");
+                            maybeLoadClass(clazz, className, fileName, results);
+                        }
+                        return FileVisitResult.CONTINUE;
+                    }
+                });
+            }
         }
         subclassCache.put(clazz, results);
         return results;
     }
 
+    @SuppressForbidden(reason = "test reads from jar")
+    private static JarInputStream jarStream(Path path) throws IOException {
+        return new JarInputStream(path.toUri().toURL().openStream());
+    }
+
+    /**
+     * Load classes from predefined packages (hack to limit the scope) and if they match the hierarchy, add them to the cache
+     */
+    private static <T> void maybeLoadClass(Class<T> clazz, String className, String location, List<Class<? extends T>> results)
```

Comment:
Maybe `attemptClassLoad()`?

Reply:
Depending on the name, the class might be loaded or not hence the 'maybe' versus attempt which suggests the loading will happen always (but it might fail).